src/validation.cpp

```diff
@@ -2005,6 +2004,32 @@ class WarningBitsConditionChecker : public AbstractThresholdConditionChecker
     }
 };
 
+namespace {
+// Helpers to ensure that GetBlockScriptFlags only returns a superset of MANDATORY_SCRIPT_VERIFY_FLAGS
+// (otherwise unminable txs could end up in the mempool)
+template<uint32_t FLAG>
+struct MandatoryFlagChecked
+{
+    static_assert((FLAG & MANDATORY_SCRIPT_VERIFY_FLAGS) == FLAG, "MandatoryFlag not included in MANDATORY_SCRIPT_VERIFY_FLAGS");
```

Comment:
Okay, I think there's a reasonable way of doing this with C++20 consteval:

```c++
class MandatoryFlags {
private:
    uint32_t value{0};
    struct CheckFlag {
        uint32_t flag;
        consteval CheckFlag(uint32_t f) {
            if ((f & MANDATORY_SCRIPT_VERIFY_FLAGS) != f) {
                throw "Block script flag not included in MANDATORY_SCRIPT_VERIFY_FLAGS";
            }
            flag = f;
        }
    };
public:
    explicit MandatoryFlags(const CheckFlag& flags) { value |= flags.flag; }
    void operator|=(const CheckFlag& flags) { value |= flags.flag; }
    void operator&=(uint32_t flags) { value &= flags; }
    operator uint32_t() const { return value; }
};
```

then declare `MandatoryFlags flags;` and you can just say `flags |= SCRIPT_VERIFY_DERSIG;` as normal, but if you give a non-mandatory flag, you get:

```
validation.cpp:2072:14: error: call to consteval function '(anonymous namespace)::MandatoryFlags::CheckFlag::CheckFlag' is not a constant expression
    flags |= SCRIPT_VERIFY_MINIMALIF;
             ^
validation.cpp:2018:17: note: subexpression not valid in a constant expression
                throw "Block script flag not included in MANDATORY_SCRIPT_VERIFY_FLAGS";
                ^
```

and saying `flags |= var` also gives you a decent error:

```
validation.cpp:2049:18: error: call to consteval function '(anonymous namespace)::MandatoryFlags::CheckFlag::CheckFlag' is not a constant expression
        flags |= var;
                 ^
validation.cpp:2049:18: note: initializer of 'var' is not a constant expression
validation.cpp:2046:20: note: declared here
        const auto var = it->second;
                   ^
```